### ISSUE TITLE
Download JSON from external APIs

### ISSUE THREAD

**Issue by gregoryjscott** [cityoffortworth/FortWorthTexas.gov#15]
Contributors can configure data used by Jekyll pages to be downloaded from external APIs so that data from outside sources can be integrated directly into Jekyll pages.

**Comment by bechatman** [comment]
I found two candidates for using this method: Bid Openings and Calendar. Seems that it may be easier to have people maintain that in the data portal then manipulate the output via liquid. We probably need to evaluate whether collections or external data would be the best method.

**Comment by bechatman** [comment]
Had a thought on this based on your idea that we just edit some yaml to make this work. Could we use a similar format to the front matter defaults structure and have a single plugin (or separate ones based on file type)? What I am envisioning is:
external-data:
- 
  remote: "http://data.source.com/api.{xml | json | csv}"
    parent: "entries"
    map:
      - variable1: "api-element1"
      - variable2: "api-element2"
    output: {true | false} // would just write 
    dest: "_collection-name" // if output is true it generates pages; if false, you give a filename and it will just store a copy  in _data
    cache: {true | false}
    cache-interval: 24 // in hours, so it would need to store an expires header of some kind.

**Comment by gregoryjscott** [comment]
This might work. The only thing that might get a little hairy is the `map:` section. If the mappings remain simple then it should be fine, but if there's a situation where you need to transform/modify the data then YAML wouldn't be a good fit.

**Comment by bechatman** [comment]
Yeah I think this is a case where you just pull the data directly. I think anything more complicated should probably built into a single-purpose plugin. The only thing we might want to sort out is built-in conversions (like date/time conversions into UTC format or something like that). Maybe the `map` element could accept two arguments: Source field and data type.

**Comment by bechatman** [comment]
Just saw something that could cause some problems on the Wufoo front, but there may be a workaround. The entries API only allows 100 entries to be pulled at a time, but it does give a number of pages available and allow you to cycle through the pages to pull entries. They also have to query strings that all you to filter the results to get them under the 100 limit.

I think that calendar is probably the best example for using the filter to only pull future dates. The purchasing end, though requires all entries to be retained, so there would need to be a loop to get the number of pages and cycle the pages. 

Also, I am not sure if this came up and I forgot it, but we could roll all these API calls into a single rake task that updates all of these data files and pushes the commit to master, then fire an email with the number of changes that need to be pushed to production (number of new entries for example). If we set it to run at Noon so it generates an email after the first round of people get here and make their changes and then at 5 p.m. so we get any changes made during the rest of the day.

**Comment by gregoryjscott** [comment]
I'm still bothered by the idea of having to commit data from the third party API into this repository to update the content. Here's my thinking.

We setup a Rake task, perhaps `rake download:roundup-calendar` that downloads the data to `_data`. You or other contributors will run it every now and then and commit the new data but that will only be for development purposes. This means your recurring local builds stay fast.

The CI server will run `rake download:roundup-calendar` before each build so that deployed versions of the site always use the most recent data. However, I don't see a need to commit these changes every time. While I do like to commit data files for one-off loads (if size is reasonable), I don't think committing data from a source that has it's own API is necessary. If for some business reason you needed to archive that data then I'd do it differently - I wouldn't mix those commits in with all the regular content based commits of this repository.

**Comment by bechatman** [comment]
Given all the issues we discussed and given some time to think of it, I think you're right. With the "on commit" build of the site, it is just easier that way. I say just move forward with your approach. FYI: I am having a couple of the admins that work with me on the site load calendar data in both of our calendar forms over the next couple of days, so we should have more sample data to work with. I also got that feature section schedule to work via CSV, and the before/after tags worked great.
